text/0000-dockerfiles.md

```diff
@@ -0,0 +1,228 @@
+* <new lifecycle phase OR platform> apply Dockerfiles to run image (could run in parallel with build image extension)
+* <new lifecycle phase OR platform> apply Dockerfiles to build image
+* restore
+* build
+* export
+
+When Dockerfiles are used to update the run image, care should be taken to ensure registry access prior to the `build` phase, to avoid long builds that fail due to incorrect credentials.
+
+### Dynamically-applied Dockerfiles
+
+A builder image may include any number of "extensions" directories in `/cnb/extensions/`.
+
+Extensions are similar to buildpacks: they have two executables: `/bin/detect` and `/bin/generate`. The interface for these executables is similar to a buildpack's `/bin/detect` and `/bin/build`.
+However, instead of a `buildpack.toml` file, extensions have a `extension.toml` file:
+```toml
+api = "<buildpack API version>"
+
+[extension]
+id = "<extension ID>"
+name = "<extension name>"
+version = "<extension version>"
+homepage = "<extension homepage>"
+description = "<extension description>"
+keywords = [ "<string>" ]
+
+[[extension.licenses]]
+type = "<string>"
+uri = "<uri>"
+```
+
+Extensions may be packaged and examined similar to buildpacks, but with analogous `pack extension` subcommands.
+
+Other pack CLI commands, such as `pack builder create`, will be extended to include support for extensions.
+
+Unlike buildpacks,
+- Extensions must not be included in a meta-buildpacks
+- Extensions must not have `order`/`group` definitions in `extension.toml`
+
+Extensions participate in the buildpack detection process, with the same UID, GID, and interface for `/bin/detect`.
+However,
+- `/bin/detect` is optional for extensions, and they are assumed to pass detection when it is not present. Just like with buildpacks, a /bin/detect that exits with a 0 exit code passes detection, and fails otherwise.
+- If an extension is missing `/bin/detect`, the extension root is treated as a pre-populated output directory (i.e., extensions can include a static build plan).
+- Extensions may only output `provides` entries to the build plan. They must not output `requires`.
+- Extensions are not included in `order` definitions (e.g., in `builder.toml`); instead, a separate `order-extensions` table should be used. The `order-extensions` table will be prepended to each group in the provided `order` (as if `order-extensions` were a composite buildpack).
+- Extensions are always `optional`.
+
+Extensions generate Dockerfiles before the regular buildpack build phase.
+To generate these Dockerfiles, the lifecycle executes the extension's `/bin/generate` executable with the same UID, GID, and interface as regular buildpacks.
+However,
+- Extensions `/bin/generate` must not write to the app directory.
+- Extensions `<layers>` directory is replaced by an `<output>` directory.
+- If an extension is missing `/bin/generate`, the extension root is treated as a pre-populated `<output>` directory.
```

Comment:
This combined with the following from above

> If an extension is missing `/bin/detect`, the extension root is treated as a pre-populated output directory (i.e., extensions can include a static build plan).

Makes the same directory the output of both detect and generate. Would it be tidier and more future proof to have two separate directories for these two purposes? 

The following is a proposed filetree for all possible files (although some of these would be mutually exclusive).
```
extension-root
├── bin
│   ├── detect
│   └── generate
├── detect
│   └── plan.toml
└── generate
    ├── Dockerfile
    ├── build.Dockerfile
    ├── build.toml
    ├── launch.toml
    └── run.Dockerfile
```

This organization might be particularly nice if we wanted to provide the same (static) options for regular buildpacks in the future.

Reply:
I had the same thought myself, but I think it would be easier to implement this if both buildpacks and extensions followed the same convention - i.e., when we get to the point of providing static options for regular buildpacks, let's implement it for extensions also.

Reply:
Updated in 3a26570